catalog_tools/utils/binning.py

```diff
@@ -0,0 +1,73 @@
+import numpy as np
+import decimal
+
+
+def normal_round_to_int(x: float) -> int:
+    """
+    Rounds a float number x to the closest integer.
+
+    Args:
+        x: decimal number that needs to be rounded
+
+    Returns:
+        Rounded value of the given number.
+    """
+
+    sign = 1 if (x >= 0) else -1
+    x = abs(x)
+
+    if x - np.floor(x) < 0.5:
+        return int(sign * np.floor(x))
+    return int(sign * np.ceil(x))
+
+
+def normal_round(x: float, n: int) -> float:
+    """
+    Rounds a float number x to n number of decimals. If the number
+    of decimals is not given, we round to an integer.
+
+    Args:
+        x: decimal number that needs to be rounded
+        n: number of decimals, optional
+
+    Returns:
+        Value rounded to the given number of decimals.
+    """
+
+    power = 10**n
+    return normal_round_to_int(x * power) / power
+
+
+def bin_to_precision(x: float, delta_m: float = 0.1) -> float:
+    """
+    Rounds a float number x to a given precision. If precision not given,
+    assumes 0.1 bin size
+
+    Args:
+        x: decimal number that needs to be rounded
+        delta_m: size of the bin, optional
+
+    Returns:
+        Value rounded to the given precision.
+    """
+
+    d = decimal.Decimal(str(delta_m))
+    decimal_places = abs(d.as_tuple().exponent)
+    return np.round(normal_round_to_int(x / delta_m) * delta_m, decimal_places)
+
+
+def bin_magnitudes(mags: list, delta_m: float = 0.1) -> list:
+    """
+    Rounds a list of float numbers to a given precision.
+    If precision not given, assumes 0.1 bin size.
+
+    Args:
+        mags: list of values that need rounding
+        delta_m: size of the bin, optional
+
+    Returns:
+        List of values rounded to the given precision.
+    """
+
+    rounded_list = [bin_to_precision(m, delta_m) for m in mags]
```

Comment:
I'm not sure about this, but it could be faster to turn the list into a numpy array and then apply the function to the array instead of doing a for loop through the list. If you have long lists this will make a difference. for loops are generally bad :)

Reply:
You're probably right, I had to make some minor changes to the functions so that they would work on arrays also - not sure if I should specify this in the input types?

Reply:
for loops aren't evil per se. Here I think two things are important to consider:
- complexity of function
- size of array

complex function: write the function to operate on array
easy function, huge array (100000(0) upwards)/executed A LOT of times: use numpy arrays or dataframes, think about how you apply the methods to the array: https://stackoverflow.com/questions/35215161/most-efficient-way-to-map-function-over-numpy-array

if neither the function is very complex nor the array is very big(<10^6) and the function isn't being executed too often, then it just simply doesnt matter and list comprehension (for loop) is a very readable and quick way to do operations.

An easy way to get a feeling for this is timing the execution time.... I often timed functions and had to notice that it simply doesn't matter